src/support/istring.cpp

```diff
@@ -17,41 +17,61 @@
 #include "istring.h"
 
 namespace wasm {
+namespace {
 
-std::string_view IString::interned(std::string_view s, bool reuse) {
-  // We need a set of string_views that can be modified in-place to minimize
-  // the number of lookups we do. Since set elements cannot normally be
-  // modified, wrap the string_views in a container that provides mutability
-  // even through a const reference.
-  struct MutStringView {
-    mutable std::string_view str;
-    MutStringView(std::string_view str) : str(str) {}
-  };
-  struct MutStringViewHash {
-    size_t operator()(const MutStringView& mut) const {
-      return std::hash<std::string_view>{}(mut.str);
-    }
-  };
-  struct MutStringViewEqual {
-    bool operator()(const MutStringView& a, const MutStringView& b) const {
-      return a.str == b.str;
-    }
-  };
-  using StringSet =
-    std::unordered_set<MutStringView, MutStringViewHash, MutStringViewEqual>;
+// We need a set of string_views that can be modified in-place to minimize
+// the number of lookups we do. Since set elements cannot normally be
+// modified, wrap the string_views in a container that provides mutability
+// even through a const reference.
+struct MutStringView {
+  mutable std::string_view str;
+  MutStringView(std::string_view str) : str(str) {}
+};
+struct MutStringViewHash {
+  size_t operator()(const MutStringView& mut) const {
+    return std::hash<std::string_view>{}(mut.str);
+  }
+};
+struct MutStringViewEqual {
+  bool operator()(const MutStringView& a, const MutStringView& b) const {
+    return a.str == b.str;
+  }
+};
+using StringSet =
+  std::unordered_set<MutStringView, MutStringViewHash, MutStringViewEqual>;
 
-  // The authoritative global set of interned string views.
+// The authoritative global set of interned string views.
+StringSet& getGlobalStrings() {
   static StringSet globalStrings;
+  return globalStrings;
+}
 
-  // The global backing store for interned strings that do not otherwise have
-  // stable addresses.
+// The global backing store for interned strings that do not otherwise have
+// stable addresses.
+std::vector<std::vector<char>>& getAllocated() {
   static std::vector<std::vector<char>> allocated;
+  return allocated;
+}
 
-  // Guards access to `globalStrings` and `allocated`.
+// Guards access to `globalStrings` and `allocated`.
+std::mutex& getMutex() {
   static std::mutex mutex;
+  return mutex;
+}
 
-  // A thread-local cache of strings to reduce contention.
+// A thread-local cache of strings to reduce contention.
+StringSet& getLocalStrings() {
   thread_local static StringSet localStrings;
+  return localStrings;
+}
+
+} // anonymous namespace
+
+std::string_view IString::interned(std::string_view s, bool reuse) {
+  auto& mutex = getMutex();
+  auto& allocated = getAllocated();
+  auto& localStrings = getLocalStrings();
+  auto& globalStrings = getGlobalStrings();
```

Comment:
This is correct but I wonder if it has any performance downside. This is called on every string, so if these function calls are not inlined that is worrying. I suppose they likely are, but I wonder if there is a way to guarantee that, by avoiding function calls like this?

(Are the function calls to avoid the global initializer issue? If so a comment in the source would be good.)

Reply:
> This is correct but I wonder if it has any performance downside. This is called on every string, so if these function calls are not inlined that is worrying. I suppose they likely are, but I wonder if there is a way to guarantee that, by avoiding function calls like this?
> 
> (Are the function calls to avoid the global initializer issue? If so a comment in the source would be good.)

Yes, the reason why I explicitly make these function calls is to guarantee the initialization of static variables, otherwise the behavior may be undefined.
For the performance downside, I think your concern is reasonable but unfortunately I don't know how to avoid this kind of function call in a deterministic way. Hope tlively could give some suggestions.

Reply:
Another way might be to put those things in a class?
```cpp
namespace {
struct StringStorage {
  StringSet globalStrings;
  std::vector<std::vector<char>> allocated;
  std::mutex mutex;

  auto intern(..);

  auto DANGER_clear();
} stringStorage;
}

std::string_view IString::interned(std::string_view s, bool reuse) {
  return stringStorage.intern(..);
}
```
That still leaves a call, but at least it is just one.

Reply:
> Another way might be to put those things in a class?
> 
> ```c++
> namespace {
> struct StringStorage {
>   StringSet globalStrings;
>   std::vector<std::vector<char>> allocated;
>   std::mutex mutex;
> 
>   auto intern(..);
> 
>   auto DANGER_clear();
> } stringStorage;
> }
> 
> std::string_view IString::interned(std::string_view s, bool reuse) {
>   return stringStorage.intern(..);
> }
> ```
> 
> That still leaves a call, but at least it is just one.

Thanks again for your patience, kripken. Unfortunately, I decide to **give up this pull request** since maintaining a name pool during the fuzzing process seems more straightforward for me, and the experimental result proves that the overall memory usage is effectively under control. Thus, I think keeping Binaryen's original implementation may be better ;).